### PR TITLE
feat(cli): add SSH support for `renku session`

### LINKED ISSUE #3289: add `renku session ssh-setup` command
Add a new subcommand `renku session ssh-setup` that
- checks that a user is logged in
- checks that a compatible ssh client is installed in a project
- (optional) checks if ssh-agent is available
- checks if a users keys are already installed for a project
  * In this case, command exits and nothing is changed, unless called with the `--yes` flag
- generates private and public keys for the user (or use existing one if called with `--existing-key=....`
- sets up the users ssh config with the jump host for that renku deployment (1)
- Adds the users public key to the projects authorized_keys file
- checks if VSCode is installed and sets it up to use that ssh connection
- informs the user that they are ready to start ssh enabled sessions and how to do so

(1): Example config for this looks like
```
Host renkulab-jumphost
  HostName renkulab.io
  Port 2022
 
Host user-project-*
  ProxyJump  renkulab-jumphost
  IdentityFile ~/.ssh/renku-user-project.key
  User jovyan
```

where `project-user-session` is the session name and `renkulab.io` is the URL of the jump host for the logged in renku instance. 
`user-project-*` has to be the proper beginning of the session name, e.g. `john-2edoe-test-2dprotected-2dbranc-*`, but without commit hash.

the config file is stored in `~/.ssh/renku/jumphost` and `~/.ssh/renku/user-project-main` with an `Include ~/.ssh/renku/` directive added to the main `~/.ssh/config` file.

When a session gets started later, it only needs an entry for the `HostName` and all settings such as key are persisted in the `user-project-*` entry.

### LINKED ISSUE #3290: add `--ssh` option to `renku session start`
blocked by https://github.com/SwissDataScienceCenter/renku-python/issues/3289

- checks that a users ssh config is correct for a project (otherwise performs the steps outline in https://github.com/SwissDataScienceCenter/renku-python/issues/3289)
- Raises an error if run with local (docker) provider.
- Starts the session with SSH turned on
- Create a new ssh config entry on the users system

SSH config entry looks like
```
Host renku-session-john-2edoe-test-2dprotected-2dbranc-4e9e126a-0
    HostName john-2edoe-test-2dprotected-2dbranc-4e9e126a-0
```
stored in `~/.ssh/renku/renku-session-john-2edoe-test-2dprotected-2dbranc-4e9e126a-0`
prints out the command to connect via SSH, e.g. `ssh renku-session-john-2edoe-test-2dprotected-2dbranc-4e9e126a-0`

### PR DESCRIPTION
**Pull request by Panaetius**
closes #3289 
closes #3290 

Adds `renku session setup-ssh` command to configure SSH locally for access to a deployment (need to be logged in).
Adds `renku session start --ssh` command that setups of keys in the project for SSH access and configures the local machine for access to a session. Also perfirms `setup-ssh` if not done already.
Adds `renku session open --ssh` command that opens an SSH connection to a session.

### CODE REVIEW

**Comment by Panaetius** [comment]
Steps to test:
- Create a project there and set the base image in the Dockerfile to `ARG RENKU_BASE_IMAGE=renku/renkulab-py:python-3.9.12-4d92ea3`
- Clone the project locally
- Install renku on your local machine from [this PR](https://github.com/SwissDataScienceCenter/renku-python/pull/3313) `pip install git+https://github.com/SwissDataScienceCenter/renku-python.git@topic/3289-ssh-session-setup`
- `renku login renku-ci-nb-1389.dev.renku.ch`
- `renku session setup-ssh` (optional, sets up SSH jump host and keys for the deployment). you can also use `-k <path to private key>` if you don't want a key generated for you and use an existing key instead
- `export RENKU_IMAGE_REGISTRY="registry.dev.renku.ch"` this is a workaround since renku session start doesn't currently work for external gitlab deployments where gitlab is not at gitlab.<host> or <host>/gitlab (which our CI deployments are). On dev or renkulab this wouldn't be necessary
- `renku session start -p renkulab --ssh` starts the session, adds your key to authorized_hosts(the first time) and prints the connection name
- Connect with VScode with the printed session name or `renku session open --ssh <session id>` for just a shell

**Review by mohammad-alisafaee — CHANGES_REQUESTED** [review]
Thanks Ralf! This is great! I couldn't create a project in the deployed env; it hangs when fetching templates.

File: renku/core/session/session.py [hunk 1]
```diff
@@ -183,21 +216,95 @@ def stop_sessions(session_provider: ISessionProvider) -> bool:
...
+    if existing_key:
+        communication.info("Linking existing keys")
+        existing_public_key = existing_key.parent / (existing_key.name + ".pub")
+
+        if not existing_key.exists() or not existing_public_key.exists():
+            raise errors.KeyNotFoundError(
+                f"Couldn't find private key '{existing_key}' or public key '{existing_public_key}'."
+            )
+
+        system_config.keyfile.unlink(missing_ok=True)
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/session/session.py:278]: it's still 6 months early to retire it

File: renku/core/session/renkulab.py [hunk 2]
```diff
@@ -148,8 +156,51 @@ def pre_start_checks(self):
...
+                    raise errors.RenkulabSessionError(
+                        "Can't run ssh session without setting up Renku SSH support. Run without '--ssh' or "
+                        "run 'renku session setup-ssh'."
+                    )
+
+            project_context.ssh_authorized_keys_path.parent.mkdir(parents=True, exist_ok=True)
+            project_context.ssh_authorized_keys_path.touch(mode=0o644, exist_ok=True)
+
+            key = system_config.public_keyfile.read_text()
+            key = f"\n{key} {project_context.repository.get_user().name}"
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/session/renkulab.py:177]: I had it without originally and that just appended it to the same line

File: renku/core/util/ssh.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    def __init__(self) -> None:
+        """Initialize class ans calculate paths."""
+        self.ssh_root: Path = Path.home() / ".ssh"
+        self.ssh_config: Path = self.ssh_root / "config"
+        self.renku_ssh_root: Path = self.ssh_root / "renku"
+
+        self.renku_ssh_root.mkdir(exist_ok=True, parents=True)
+        self.ssh_config.touch(mode=0o644, exist_ok=True)
+
+        self.renku_host: str = str(urllib.parse.urlparse(get_renku_url()).hostname)
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/util/ssh.py:65]: That's mostly to keep mypy happy, it's always a string here but the calculated type is `str | Unknown | None`. I'll use a cast instead

File: renku/core/util/ssh.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    """Class to manage system SSH config."""
+
+    def __init__(self) -> None:
+        """Initialize class ans calculate paths."""
+        self.ssh_root: Path = Path.home() / ".ssh"
+        self.ssh_config: Path = self.ssh_root / "config"
+        self.renku_ssh_root: Path = self.ssh_root / "renku"
+
+        self.renku_ssh_root.mkdir(exist_ok=True, parents=True)
+        self.ssh_config.touch(mode=0o644, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/util/ssh.py:63]: it is a file `~/.ssh/config`. I thin for SSH itself both 644 and 600 are ok (it was 644 by default on my system), anything more and it'll complain.

I'll change everything to follow https://superuser.com/a/1559867

File: renku/core/session/renkulab.py [hunk 5]
```diff
@@ -148,8 +156,51 @@ def pre_start_checks(self):
...
+
+            if key in project_context.ssh_authorized_keys_path.read_text():
+                return
+
+            communication.info("Adding SSH public key to project.")
+            with project_context.ssh_authorized_keys_path.open("at") as f:
+                f.writelines(key)
+
+            repository.add(project_context.ssh_authorized_keys_path)
+            repository.commit("Add SSH public key.")
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/session/renkulab.py:187]: In the session (remote), we have an SSH server running and this file tells it which keys are allowed to connect. So the keys are local but the public key gets added to the project here so it's available remotely/in the session.

Once we have a proper ssh-proxy, we can hopefully have that manage keys and not need to check them into the project.

**Comment by Panaetius** [comment]
> I couldn't create a project in the deployed env; it hangs when fetching templates.
That weird. I guess you could always just create the project on Dev as well and it'd show up here.
Or maybe it'll work now that I'm redeploying it with the new commit?

File: renku/ui/cli/session.py [hunk 6]
```diff
@@ -279,8 +333,26 @@ def stop(session_name, stop_all, provider):
...
+@session.command("ssh-setup")
+@click.option(
+    "existing_key",
+    "-k",
+    "--existing-key",
+    type=click.Path(exists=True, dir_okay=False),
+    metavar="<private key file>",
+    help="Existing private key to use.",
+)
+@click.option("--force", is_flag=True, help="Overwrite existing keys/config.")
```
Comment (mohammad-alisafaee) [inline renku/ui/cli/session.py:353]: One more thing, I think `--yes` makes more sense in this context since we show a confirmation prompt to user.

Comment (Panaetius) [reply renku/ui/cli/session.py:353]: I had `--yes` and it was voted against in favor of `--force` by a 3:1 majority ... :man_shrugging:

**Review by mohammad-alisafaee — APPROVED** [review]

**Pull request merged**